src/Trains.NET.Rendering/LayerRenderer/Bases/LayoutRenderer.cs

```diff
@@ -40,40 +44,85 @@ public void Render(ICanvas canvas, int width, int height, IPixelMapper pixelMapp
 
                 canvas.Translate(x, y);
 
+                float heightScale = 1;
+                if (this.IsScaledByHeight)
+                {
+                    heightScale = GetScaleForTerrainHeight(entity.Column, entity.Row);
+                }
+
                 if(_renderer is ICachableRenderer<T> cachableRenderer)
                 {
                     canvas.ClipRect(new Rectangle(0, 0, pixelMapper.CellSize, pixelMapper.CellSize), false);
 
                     string key = cachableRenderer.GetCacheKey(entity);
 
-                    if (!_cache.TryGetValue(key, out IImage cachedImage))
+                    if (!_cache.TryGetValue((key, heightScale), out IImage cachedImage))
                     {
                         using IImageCanvas imageCanvas = _imageFactory.CreateImageCanvas(pixelMapper.CellSize, pixelMapper.CellSize);
 
                         float scale = pixelMapper.CellSize / 100.0f;
-
                         imageCanvas.Canvas.Scale(scale, scale);
 
+                        if (heightScale < 1)
+                        {
+                            imageCanvas.Canvas.Scale(heightScale, heightScale, 50, 50);
+                        }
+
                         _renderer.Render(imageCanvas.Canvas, entity);
 
                         cachedImage = imageCanvas.Render();
 
-                        _cache[key] = cachedImage;
+                        _cache[(key, heightScale)] = cachedImage;
                     }
 
                     canvas.DrawImage(cachedImage, 0, 0);
                 }
                 else
                 {
                     float scale = pixelMapper.CellSize / 100.0f;
-
                     canvas.Scale(scale, scale);
+                    if (heightScale < 1)
+                    {
+                        canvas.Scale(heightScale, heightScale, 50, 50);
+                    }
 
                     _renderer.Render(canvas, entity);
                 }
 
                 canvas.Restore();
             }
         }
+
+        private float GetScaleForTerrainHeight(int column, int row)
```

Comment:
I'm surprised you think `GetScaleForTerrainHeight` belongs on the base layer renderer.